EIPS/eip-3520.md

```diff
@@ -0,0 +1,93 @@
+---
+eip: 3520
+title: Transaction Destination Opcode
+author: Alex Papageorgiou (@alex-ppg)
+discussions-to: https://ethereum-magicians.org/t/eip-3520-transaction-destination-opcode/6058
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-04-16
+requires: 3508
+---
+
+## Simple Summary
+
+Provide access to the original recipient of a transaction.
+
+## Abstract
+
+This EIP introduces the following EVM instruction: `ENTRYPOINT`.
+
+This instruction is meant to provide access to the original recipient of the transaction, the `to` address, enabling new ways of introspection to be applied in conjunction with [EIP-3508](./eip-3508.md).
+
+## Motivation
+
+It is undeniable that smart contracts are becoming more interconnected than ever. Up until this point, smart contracts have entirely relied on compliant interfaces and introspection to introduce a new step in the call chain of a complex multi-contract interaction. However, this presents a forwards-only approach which limits the types of interactions that can manifest.
+
+The purpose of this EIP is to provide a way via which a contract is able to identify the entry-point of a transaction on the blockchain and deduce what was the original intention of the transaction by applying introspection on the original transaction data itself.
+
+This EIP enables the development of new types of smart contracts as it can open new pathways for [EIP-721](./eip-721) NFTs and [EIP-20](./eip-20) tokens to detect which action their transaction is part of, such as detecting a liquidity provision to a decentralized exchange or a loan within a collateralized lending protocol.
+
+## Specification
+
+### ENTRYPOINT (`0x4a`)
+
+The `ENTRYPOINT` instruction uses 0 stack arguments and pushes the original `to` member of the transaction onto the stack. The address yielded by the instruction is a 160-bit value padded to 256-bits. The operation costs `G_base` to execute, similarly to `ORIGIN` (`0x32`).
+
+### AUTHCALL (`0xf7`)
+
+The `ENTRYPOINT` opcode by itself re-introduces discrimination into the system as it indirectly allows one to evaluate whether the smart contract code being executed is done so by an EOA by validating that `ENTRYPOINT == ADDRESS` where `ADDRESS` (`0x30`) retrieves the currently executing account address.
+
+In detail, the address returned by the `ENTRYPOINT` opcode will be equivalent to the `to` address parameter specified in the nearest `AUTHCALL` up the stack.  If there is no `AUTHCALL` in the stack then `ENTRYPOINT` will retrieve the original transaction's `to` field.
```

Comment:
Suggestion: This feels like it should be in the same paragraph/sentence as the definition of the `ENTRYPOINT` opcode, since it is a critical part of `ENTRYPOINT`'s definition.  As it is structured currently, this makes it seem like a separate unrelated thing but it very much is not.